tests/framework/behaviors/CacheableBehaviorTest.php

```diff
@@ -0,0 +1,221 @@
+
+    /**
+     * @inheritdoc
+     */
+    protected function setUp()
+    {
+        $this->initializeApplicationMock();
+        $this->initializeWidgetMocks();
+    }
+
+    public function testDefaultCache()
+    {
+        $this->assertEquals('cache', $this->simpleWidget->cache);
+    }
+
+    public function testDefaultDuration()
+    {
+        $this->assertEquals(60, $this->simpleWidget->cacheDuration);
+    }
+
+    public function testDefaultDependency()
+    {
+        $this->assertNull($this->simpleWidget->cacheDependency);
+    }
+
+    public function testDefaultKey()
+    {
+        $this->assertNull($this->simpleWidget->cacheKey);
+    }
+
+    public function testDefaultKeyVariations()
+    {
+        $this->assertEquals([], $this->simpleWidget->cacheKeyVariations);
+    }
+
+    public function testDefaultEnabled()
+    {
+        $this->assertTrue($this->simpleWidget->cacheEnabled);
+    }
+
+    public function testWidgetIsRunWhenCacheIsEmpty()
+    {
+        $this
+            ->simpleWidget
+            ->expects($this->once())
+            ->method('run')
+        ;
+
+        $contents = $this->simpleWidget->test();
+        $this->assertEquals('contents', $contents);
+    }
+
+    public function testWidgetIsNotRunWhenCacheIsNotEmpty()
+    {
+        $this->simpleWidget->cacheDuration = 0;
+        $this
+            ->simpleWidget
+            ->expects($this->once())
+            ->method('run')
+        ;
+
+        for ($counter = 0; $counter <= 42; $counter++) {
+            $this->assertEquals('contents', $this->simpleWidget->test());
+        }
+    }
+
+    public function testDynamicContent()
+    {
+        $this->dynamicWidget->cacheDuration = 0;
+        $this
+            ->dynamicWidget
+            ->expects($this->once())
+            ->method('run')
+        ;
+
+        for ($counter = 0; $counter <= 42; $counter++) {
+            $expectedContents = sprintf('<div>dynamic contents: %d</div>', $counter);
+            $this->assertEquals($expectedContents, $this->dynamicWidget->test());
+        }
+    }
+
+    /**
+     * Initializes a mock application.
+     *
+     */
+    private function initializeApplicationMock()
+    {
+        $this->mockApplication([
+            'components' => [
+                'cache' => [
+                    'class' => '\yii\caching\ArrayCache',
+                ],
+            ],
+            'params' => [
+                // Counter for dynamic contents testing.
+                'counter' => 0,
+            ],
+        ]);
+    }
+
+    /**
+     * Initializes mock widgets.
+     *
+     */
+    private function initializeWidgetMocks()
+    {
+        $this->simpleWidget = $this->getWidgetMock(SimpleCacheableWidget::className());
+        $this->dynamicWidget = $this->getWidgetMock(DynamicCacheableWidget::className());
+    }
+
+    /**
+     * Returns a widget mock.
+     *
+     */
+    private function getWidgetMock($widgetClass)
+    {
+        $widgetMock = $this
```

Comment:
@Kolyunya As I said, this is very subjective. I expect that to be covered in the docs to prevent disputes and disagreements. Since there is no such info there let's wait for others contributors' opinions. You can leave your version if you think it's better.